### DISCUSSION TITLE
Dynamic quadratic curves and point transformations

### DISCUSSION THREAD

**Discussion by SupremeDeity** [fabricjs/fabric.js#10183] [Q&A]
Hello!
I am trying to create a curve similar to: [Fabric Quadratic Curve demo](http://fabricjs.com/quadratic-curve) however there are several differences. Mainly that my line can be moved like an object and i need to pass it through roughjs. I have created a control point in the middle for now. The questions i have are:

1) Is there an easy way or efficient way to convert ScenePoint to a point relative to the object? For example my control point in the center is initially at 0,0, but on moving that control point i need to convert it to the displacement between the two points. 
2) Are there some resources or something i can look at to understand what all these transformations fabric is doing? Like this which i copied to make it look like the arrow is being drawn from the mouse point and not the center, and also whatever scenepoint, viewpoint and the transformations using the `Point` class:
```js
  const relativeCenter = this.getRelativeCenterPoint();
        const constraint = this.translateToOriginPoint(
            relativeCenter,
            originX,
            originY
        );
```
Here is the current demo (use the line shape): [My Curved Line demo](https://kaku-git-curved-lines-supremedeitys-projects.vercel.app)
Here is the code: [Custom quadratic line code](https://github.com/SupremeDeity/kaku/blob/curved-lines/utils/roughLine.js)
What i want to achieve is similar to excallidraw: 

https://github.com/user-attachments/assets/35413e6e-4bde-4f05-94ed-131e15415ca3

**Comment by SupremeDeity** [comment]
Ok so i had the demo url wrong and it was protected so others wouldn't be able to see but i have fixed that. 
I tried using a easier approach to get relative position for the middlepoint:
```js
this.controls.bendControls = new fabric.Control({
            cursorStyle: "pointer",
            x: (this.midpoint[0]), y: this.midpoint[1],
            actionHandler: (eventData) => {
                const pointer = this.canvas.getScenePoint(eventData);
                const centerPoint = this.getCenterPoint();
                const deltaX = pointer.x - centerPoint.x;
                const deltaY = pointer.y - centerPoint.y;

                this.midpoint = [deltaX, deltaY];
                // console.log(displacement)
                this.update();
                this.canvas.requestRenderAll();
                this.canvas.fire("object:modified");
            },
        });
```

However this doesn't seem to work very well. My original approach in the link i provided in the message *does* work but i wanted to see if a much easier approach was available. Other than that there's a bit of issues of how i cant rely on left and top because changing position messes with it.

**Reply by asturur** [reply]
There is a tool in fabric to do this embedded, does not work for you? the path controls points introduced recently

**Reply by SupremeDeity** [reply]
> There is a tool in fabric to do this embedded, does not work for you? the path controls points introduced recently

Maybe i dont know of the thing. Is it fabric v6? I will google it. The main point is i need to pass it to rough.js so i either need points or the path string. 

Edit: Looked into it a bit and i didnt get what you meant. Did you meant just using the fabric.Path?

**Reply by SupremeDeity** [reply]
@asturur sorry, pinging just in case you didnt see this.

**Reply by SupremeDeity** [reply]
Bump! really want to know what this native tool is

**Reply by asturur** [reply]
Sorry i have been on paternity leave quite busy here.

There is this tool to add control points to curves in fabricJS:
https://github.com/fabricjs/fabric.js/pull/9998

I wanted to add a guide but i have no time.
Check this sample was added as a quick test
https://github.com/fabricjs/fabric.js/pull/9998/files#diff-70f24eb979856f3b260b75b17ccd49fa9cbc218efdedaf22d8d934cdb9854ddf

and use your quadratic curve there, check if is what you need.

**Reply by SupremeDeity** [reply]
@asturur So i checked it out. The problem i have is that the actual drawing is done by rough, hence i cant really be adding `fabric.Path` itself to the fabric canvas. I dont have the time right now and its quite late at night so my attempt at getting this to work is quite funky.

https://github.com/SupremeDeity/kaku/blob/22e9f23ebcc24721ce2e5a57ef15601eecbd14b7/utils/roughLine.js#L50

Its just changing the parent type from `fabric.FabricObject` -> `fabric.Path`, then i just initialize with an `undefined` path, and set the actual path in the `_updateRoughLine()` function. This gets the control points to show atleast, but ofcourse there's too many issues like the shape not being updated correctly, which i believe is because its getting overwritten each time. I guess i'll update here whenever i get the time (most probably on the weekends), and ask if i still cant get this to work somehow.

Here's the deployment url for this branch: [Test deployment](kaku-git-curved-lines-supremedeitys-projects.vercel.app)

**Reply by SupremeDeity** [reply]
@asturur So i got the time to work more on this today. I haven't gotten to the quadratic curve part yet but from what i am seeing for some reason the `Path.path` is not getting updated. The controls are changing as they should however. This causes me to have to manually calculate the path, which is inefficient and also the dumb way i am doing it causes **both** points to change.

Here's a preview of what is happening: [Test Deployment](https://kaku-git-curved-lines-supremedeitys-projects.vercel.app).

Here's the current code, specifically line 80-91:
https://github.com/SupremeDeity/kaku/blob/909b2e88a3e8521254597a9fb36f9a3a6d798088/utils/roughLine.js

Can you tell me if this intentional ( the path not getting recalculated) ? is there a way around it so i dont have to do this manual calculation. I thought it would automatically change since im passing `this` to the createControl.

edit: I found a reply of yours saying to use `util.transformPath`, and i tried using it together with the path + calcOwnMatrix and the path's offset, it just multiplies everything to absurd coordinates.

**Reply by asturur** [reply]
I ensure you i do not understand what are you doing or what i should do in that test app to understand.
What path is not being recalculated? What happens on screen?

**Reply by SupremeDeity** [reply]
> I ensure you i do not understand what are you doing or what i should do in that test app to understand.
> What path is not being recalculated? What happens on screen?

Ah sorry 😅. I am using path controls that you suggested. I thought that would would update the `Path.path`(the parsed path inside the fabric.Path), but it apparently doesn't. Is there a way to make it do so? `fabric.util.transformPath` doesn't work well it seems. I was just passing it the path's offset and the result of calcOwnMatrix.

**Reply by asturur** [reply]
It definetely should do that indeed. It should update Path.path, yes.

**Reply by SupremeDeity** [reply]
> It definetely should do that indeed. It should update Path.path, yes.

Very weird, idk why its happening with me then, cause the control points are updating correctly, but the Path.path doesn't change. I doubt its because im subclassing fabric.Path for this usecase.

**Reply by asturur** [reply]
i can't find the quadratic curve on the links you give, or i would checked. there are only straight lines

**Reply by SupremeDeity** [reply]
> i can't find the quadratic curve on the links you give, or i would checked. there are only straight lines

I have not gotten to quadratic curves, adding a quadratic curve control point would be just adding a `Q 0 0 0 0` i think? 

Currently i only have straight lines, when i use the control points from the `utils`, they dont seem to update the `Path.path`

**Comment by SupremeDeity** [comment]
No worries, i have been busy too anyways thanks for your help, I'll look
more into it a bit later and get back to you

**Reply by asturur** [reply]
If there is a type issue we can change the util to take an object that has a path data and not necessarly a path. It can be change to an interface
